File: 동적 계획법/1463/1463.c
#include <stdio.h>
#define MAX 1000001

int main()
{
   int arr[MAX] = {0};
   int n, i;
   scanf("%d", &n);
   arr[1] = 0;

   for (i = 1; i < n + 1; ++i)
   {
      if (i * 3 < n + 1 && (arr[i * 3] == 0 || arr[i * 3] > arr[i] + 1))
         arr[i * 3] = arr[i] + 1;
      if (i * 2 < n + 1 && (arr[i * 2] == 0 || arr[i * 2] > arr[i] + 1))
         arr[i * 2] = arr[i] + 1;
      if (i < n && (arr[i + 1] == 0 || arr[i + 1] > arr[i] + 1))
         arr[i + 1] = arr[i] + 1;
   }

   printf("%d", arr[n]);

   return 0;
}

File: 동적 계획법/1463/1463.py
n = int(input())
arr = [999999999] * (n + 1)
arr[1] = 0

for i in range(1, n + 1):
   if i * 3 < n + 1 and arr[i * 3] > arr[i] + 1:
      arr[i * 3] = arr[i] + 1
   if i * 2 < n + 1 and arr[i * 2] > arr[i] + 1:
      arr[i * 2] = arr[i] + 1
   if i < n and arr[i + 1] > arr[i] + 1:
      arr[i + 1] = arr[i] + 1
   
print(arr[n])